### ISSUE TITLE
connectTCP should re-resolve domain names on each connection attempt

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1822]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#1822|
|Type|enhancement|
|Created|2006-06-14 21:38:49Z|
|Branch|https://github.com/twisted/twisted/tree/resolution-docs-1822|

This is helpful for connecting to places that do round-robin-DNS or failover based on DNS change, or even for long running processes dealing with regular DNS change.

Even if it already does so, we should document this behavior and add test.

<details><summary>Searchable metadata</summary>

```
trac-id__1822 1822
type__enhancement enhancement
reporter__itamarst itamarst
priority__low low
milestone__ 
branch__branches_resolution_docs_1822 branches/resolution-docs-1822
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1150321129000000 1150321129000000
changetime__1390148259000000 1390148259000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

It should probably re-resolve, but then recommend that if you need to make multiple connection attempts, you resolve and cache the IP somewhere yourself.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

It should respect normal DNS caching rules.  It may invoke the resolver before each connection attempt, but that resolver should be free to give back a cached result.

**Comment by twisted-trac** [comment]
|<img alt="SeanMcQ's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|SeanMcQ set owner to SeanMcQ|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="SeanMcQ's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|SeanMcQ commented|
|-|-|

Replying to [exarkun](#note_2):
> It should respect normal DNS caching rules.  It may invoke the resolver before each connection attempt, but that resolver should be free to give back a cached result.

Per RFC1123 section 6.1.3.1 (A) resolver implementations MUST implement caching and the cache MUST timeout.
(ftp://ftp.rfc-editor.org/in-notes/rfc1123.txt)

**Comment by twisted-trac** [comment]
|<img alt="SeanMcQ's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|SeanMcQ set owner to @itamarst|
|-|-|

Can you please provide more information.  It appears all DNS lookup paths implement caching with a timeout or defer it to the operating system to cache which is per spec.  The names library implements an empty cache, which technically meets the spec and is not used in the normal tcp connection path.

If you could provide more details about the expected behavior and where it deviates it would be helpful.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

I agree with SeanMcQ's last comment.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [41355]) Branching to 'resolution-docs-1822'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

I added some minor docs.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_11"></a>|@jonathanj set owner to @exarkun|
|-|-|

This looks good to me. Thanks for cleaning up the whitespace in the rest of the document.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_12"></a>|@hynek commented|
|-|-|

May I suggest reflowing lines where whitespace got killed around (such as 26ff)? Due to the empty lines there’s churn anyway…

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [41360]) Semantic linefeeds for lines that are changing anyway.

refs [#1822](https://github.com/twisted/twisted/issues/1822)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [41361]) Merge resolution-docs-1822

Author: exarkun
Reviewer: jonathanj, hynek
Fixes: [#1822](https://github.com/twisted/twisted/issues/1822)

Add some documentation to the client howto about how name resolution works with
`connectTCP`.
